lib/puppet/file_serving/http_metadata.rb

```diff
@@ -0,0 +1,29 @@
+require 'puppet/file_serving/metadata'
+
+# Simplified metadata representation, suitable for the information
+# that is available from HTTP headers.
+class Puppet::FileServing::HttpMetadata < Puppet::FileServing::Metadata
+
+  def initialize(http_response, path = '/dev/null')
+    super(path)
+
+    # ignore options that do not apply to HTTP metadata
+    @owner = @group = @mode = nil
+
+    if checksum = http_response['content-md5']
+      # convert base64 digest to hex
+      checksum = checksum.unpack("m0").first.unpack("H*").first
+      @checksum_type = 'md5'
+      @checksum = "{md5}#{checksum}"
+    elsif mtime = DateTime.httpdate(http_response['last-modified']).to_time
```

Comment:
This check fails if you try to source a directory-like page, such as http://nightlies.puppetlabs.com/facter-latest/repo_configs/rpm/. `http_response['last-modified']` returns `nil`, which causes `DateTime` throw an exception.

Once you check whether `http_response['last-modified']` is `nil`, you find that `PuppetError` is not defined, so spec testing isn't hitting it.